File: examples/02_movement.py
"""02_movement.py - Movement and Physics

Two modes (press TAB to switch):
  Mode A: Arrow keys move a box around the screen.
  Mode B: A bouncing ball with a fading trail.
Learn: is_key_down, fill_rect, fill_circle, draw_circle, draw_printf,
       get_fps, get_delta_time, float physics, wall collision
"""
import gameui as g


def main():
    game = g.GameLib()
    game.open(640, 480, "02 - Movement and Physics", True)

    # --- Mode A: Keyboard controlled box ---
    box_x, box_y = 310.0, 230.0
    box_size = 20
    box_speed = 200.0  # pixels per second

    # --- Mode B: Bouncing ball ---
    ball_x, ball_y = 320.0, 240.0
    ball_vx, ball_vy = 240.0, 180.0  # pixels per second
    ball_r = 20

    # Trail for bouncing ball
    trail_x: list[float] = []
    trail_y: list[float] = []

    show_ball = False  # False = mode A (box), True = mode B (ball)

    while not game.is_closed():
        if game.is_key_pressed(g.KEY_ESCAPE):
            break
        if game.is_key_pressed(g.KEY_TAB):
            show_ball = not show_ball

        dt = game.get_delta_time()
        if dt > 0.05:
            dt = 0.05  # prevent first-frame jump

        if not show_ball:
            # --- Mode A: Keyboard Control ---
            if game.is_key_down(g.KEY_LEFT):
                box_x -= box_speed * dt
            if game.is_key_down(g.KEY_RIGHT):
                box_x += box_speed * dt
            if game.is_key_down(g.KEY_UP):
                box_y -= box_speed * dt
            if game.is_key_down(g.KEY_DOWN):
                box_y += box_speed * dt

            # Keep inside window bounds
            box_x = max(0.0, min(box_x, game.get_width() - box_size))
            box_y = max(0.0, min(box_y, game.get_height() - box_size))
        else:
            # --- Mode B: Bouncing Ball ---
            ball_x += ball_vx * dt
            ball_y += ball_vy * dt

            if ball_x - ball_r < 0:
                ball_x = float(ball_r)
                ball_vx = -ball_vx
            if ball_x + ball_r > game.get_width():
                ball_x = float(game.get_width() - ball_r)
                ball_vx = -ball_vx
            if ball_y - ball_r < 0:
                ball_y = float(ball_r)
                ball_vy = -ball_vy
            if ball_y + ball_r > game.get_height():
                ball_y = float(game.get_height() - ball_r)
                ball_vy = -ball_vy

            # Record trail
            trail_x.append(ball_x)
            trail_y.append(ball_y)
            if len(trail_x) > 64:
                trail_x.pop(0)
                trail_y.pop(0)

        # --- Drawing ---
        game.clear(g.COLOR_BLACK)

        if not show_ball:
            # Mode A drawing
            game.fill_rect(int(box_x), int(box_y), box_size, box_size, g.COLOR_CYAN)
            game.draw_text(10, 10, "Mode A: Arrow keys to move", g.COLOR_WHITE)
            game.draw_printf(10, 25, g.COLOR_GRAY, f"Position: {box_x:.0f}, {box_y:.0f}")
        else:
            # Mode B drawing: trail
            for i, (tx, ty) in enumerate(zip(trail_x, trail_y)):
                brightness = min(40 + i * 3, 255)
                c = g.COLOR_RGB(brightness, 0, 0)
                tr = 2 + i * ball_r // 64
                game.fill_circle(int(tx), int(ty), tr, c)
            # Ball
            game.fill_circle(int(ball_x), int(ball_y), ball_r, g.COLOR_RED)
            game.draw_circle(int(ball_x), int(ball_y), ball_r, g.COLOR_WHITE)
            game.draw_text(10, 10, "Mode B: Bouncing Ball", g.COLOR_WHITE)
            game.draw_printf(10, 25, g.COLOR_GRAY,
                             f"Ball: {ball_x:.0f}, {ball_y:.0f}  Speed: {ball_vx:.0f}, {ball_vy:.0f}")

        # Common HUD
        game.draw_printf(10, 460, g.COLOR_LIGHT_GRAY,
                         f"FPS: {game.get_fps():.0f}  |  TAB: Switch Mode  |  ESC: Quit")

        game.update()
        game.wait_frame(60)


if __name__ == "__main__":
    main()

File: clib/examples/02_movement.cpp
// 02_movement.cpp - Movement and Physics
//
// Two modes (press TAB to switch):
//   Mode A: Arrow keys move a box around the screen.
//   Mode B: A bouncing ball with a fading trail.
// Learn: IsKeyDown, FillRect, FillCircle, DrawCircle, DrawPrintf,
//        GetFPS, GetDeltaTime, float physics, wall collision
//
// Win32: g++ -o 02_movement.exe 02_movement.cpp -mwindows
// SDL:   g++ -DUSE_SDL -std=c++11 -O2 -o 02_movement.exe 02_movement.cpp -lSDL2

#if defined(_WIN32) && !defined(USE_SDL)
#include "../GameLib.h"
#else
#include "../GameLib.SDL.h"
#endif

int main()
{
    GameLib game;
    game.Open(640, 480, "02 - Movement and Physics", true);

    // --- Mode A: Keyboard controlled box ---
    float boxX = 310.0f, boxY = 230.0f;
    int boxSize = 20;
    float boxSpeed = 200.0f; // pixels per second

    // --- Mode B: Bouncing ball ---
    float ballX = 320.0f, ballY = 240.0f;
    float ballVX = 240.0f, ballVY = 180.0f; // pixels per second
    int ballR = 20;

    // Trail for bouncing ball
    float trailX[64], trailY[64];
    int trailCount = 0;

    bool showBall = false; // false = mode A (box), true = mode B (ball)

    while (!game.IsClosed()) {
        if (game.IsKeyPressed(KEY_ESCAPE)) break;
        if (game.IsKeyPressed(KEY_TAB)) showBall = !showBall;

        float dt = (float)game.GetDeltaTime();
        if (dt > 0.05f) dt = 0.05f; // prevent first-frame jump

        if (!showBall) {
            // --- Mode A: Keyboard Control ---
            if (game.IsKeyDown(KEY_LEFT))  boxX -= boxSpeed * dt;
            if (game.IsKeyDown(KEY_RIGHT)) boxX += boxSpeed * dt;
            if (game.IsKeyDown(KEY_UP))    boxY -= boxSpeed * dt;
            if (game.IsKeyDown(KEY_DOWN))  boxY += boxSpeed * dt;

            // Keep inside window bounds
            if (boxX < 0) boxX = 0;
            if (boxY < 0) boxY = 0;
            if (boxX + boxSize > game.GetWidth())  boxX = (float)(game.GetWidth() - boxSize);
            if (boxY + boxSize > game.GetHeight()) boxY = (float)(game.GetHeight() - boxSize);
        } else {
            // --- Mode B: Bouncing Ball ---
            ballX += ballVX * dt;
            ballY += ballVY * dt;

            if (ballX - ballR < 0)                { ballX = (float)ballR; ballVX = -ballVX; }
            if (ballX + ballR > game.GetWidth())   { ballX = (float)(game.GetWidth() - ballR); ballVX = -ballVX; }
            if (ballY - ballR < 0)                { ballY = (float)ballR; ballVY = -ballVY; }
            if (ballY + ballR > game.GetHeight()) { ballY = (float)(game.GetHeight() - ballR); ballVY = -ballVY; }

            // Record trail
            if (trailCount < 64) {
                trailX[trailCount] = ballX;
                trailY[trailCount] = ballY;
                trailCount++;
            } else {
                for (int i = 0; i < 63; i++) {
                    trailX[i] = trailX[i + 1];
                    trailY[i] = trailY[i + 1];
                }
                trailX[63] = ballX;
                trailY[63] = ballY;
            }
        }

        // --- Drawing ---
        game.Clear(COLOR_BLACK);

        if (!showBall) {
            // Mode A drawing
            game.FillRect((int)boxX, (int)boxY, boxSize, boxSize, COLOR_CYAN);
            game.DrawText(10, 10, "Mode A: Arrow keys to move", COLOR_WHITE);
            game.DrawPrintf(10, 25, COLOR_GRAY, "Position: %.0f, %.0f", boxX, boxY);
        } else {
            // Mode B drawing: trail
            for (int i = 0; i < trailCount; i++) {
                int brightness = 40 + i * 3;
                if (brightness > 255) brightness = 255;
                uint32_t c = COLOR_RGB(brightness, 0, 0);
                int tr = 2 + i * ballR / 64;
                game.FillCircle((int)trailX[i], (int)trailY[i], tr, c);
            }
            // Ball
            game.FillCircle((int)ballX, (int)ballY, ballR, COLOR_RED);
            game.DrawCircle((int)ballX, (int)ballY, ballR, COLOR_WHITE);
            game.DrawText(10, 10, "Mode B: Bouncing Ball", COLOR_WHITE);
            game.DrawPrintf(10, 25, COLOR_GRAY, "Ball: %.0f, %.0f  Speed: %.0f, %.0f", ballX, ballY, ballVX, ballVY);
        }

        // Common HUD
        game.DrawPrintf(10, 460, COLOR_LIGHT_GRAY, "FPS: %.0f  |  TAB: Switch Mode  |  ESC: Quit", game.GetFPS());

        game.Update();
        game.WaitFrame(60);
    }
    return 0;
}
